prql-elixir/native/prql/src/lib.rs

```diff
@@ -0,0 +1,154 @@
+    } else if a == atoms::hive() {
+        D::Hive
+    } else if a == atoms::mssql() {
+        D::MsSql
+    } else if a == atoms::mysql() {
+        D::MySql
+    } else if a == atoms::postgres() {
+        D::PostgreSql
+    } else if a == atoms::sqllite() {
+        D::SQLite
+    } else if a == atoms::snowflake() {
+        D::Snowflake
+    } else {
+        D::Generic
+    }
+}
+
+impl From<CompileOptions> for prql_compiler::sql::Options {
+    /// Get `prql_compiler::sql::Options` options from `CompileOptions`
+    fn from(o: CompileOptions) -> Self {
+        prql_compiler::sql::Options {
+            format: o.format,
+            dialect: Some(dialect_from_atom(o.dialect)),
+            signature_comment: o.signature_comment,
+        }
+    }
+}
+
+#[derive(Clone, NifStruct, Debug)]
+#[module = "PRQL.Native.CompileOptions"]
+pub struct CompileOptions {
+    /// Pass generated SQL string trough a formatter that splits it
+    /// into multiple lines and prettifies indentation and spacing.
+    ///
+    /// Defaults to true.
+    pub format: bool,
+
+    /// Target dialect you want to compile for.
+    ///
+    /// Because PRQL compiles to a subset of SQL, not all SQL features are
+    /// required for PRQL. This means that generic dialect may work with most
+    /// databases.
+    ///
+    /// If something does not work in dialect you need, please report it at
+    /// GitHub issues.
+    ///
+    /// If None is used, `sql_dialect` flag from query definition is used.
+    /// If it does not exist, [Dialect::Generic] is used.
+    pub dialect: Atom,
+
+    /// Emits the compiler signature as a comment after generated SQL
+    ///
+    /// Defaults to true.
+    pub signature_comment: bool,
+}
+
+#[derive(NifTuple)]
+pub struct Response {
+    /// status atom `:ok` or `:error`
+    status: Atom,
+
+    /// result string
+    result: String,
+}
+
+#[rustler::nif]
+/// compile a prql query into sql
+pub fn compile(prql_query: &str, options: CompileOptions) -> NifResult<Response> {
+    let result = prql_compiler::compile(prql_query, Some(options.into()));
+    to_result_tuple(result)
+}
+
+#[rustler::nif]
+/// convert a prql query into PL AST
+pub fn prql_to_pl(prql_query: &str) -> NifResult<Response> {
+    to_result_tuple(
+        Ok(prql_query)
+            .and_then(prql_compiler::prql_to_pl)
+            .and_then(prql_compiler::json::from_pl),
+    )
+}
+
+#[rustler::nif]
+/// Convert PL AST into RQ
+pub fn pl_to_rq(pl_json: &str) -> NifResult<Response> {
+    to_result_tuple(
+        Ok(pl_json)
+            .and_then(prql_compiler::json::to_pl)
+            .and_then(prql_compiler::pl_to_rq)
+            .and_then(prql_compiler::json::from_rq),
+    )
+}
+
+#[rustler::nif]
+/// Convert RQ to SQL
+pub fn rq_to_sql(rq_json: &str) -> NifResult<Response> {
+    to_result_tuple(
+        Ok(rq_json)
+            .and_then(prql_compiler::json::to_rq)
+            .and_then(|x| prql_compiler::rq_to_sql(x, None)),
+    )
+}
```

Comment:
Can we do the same thing with these functions that we did with `compile`? Maybe they can all be one line each...

Reply:
Sure, will do

Reply:
@max-sixty 

```rust
pub fn prql_to_pl(prql_query: &str) -> NifResult<Response> {
    let pl = prql_compiler::prql_to_pl(prql_query);
    let result = prql_compiler::json::from_pl(Some(pl));
    to_result_tuple(result)
    // to_result_tuple(
    //     Ok(prql_query)
    //         .and_then(prql_compiler::prql_to_pl)
    //         .and_then(prql_compiler::json::from_pl),
    // )
}
```

like this?

Reply:
As per discussion in DC lets do this in a later phase